ERCS/erc-7679.md

```diff
@@ -0,0 +1,223 @@
+---
+eip: 7679
+title: UserOperation Builder
+description: Construct UserOperations without being coupled with account-specific logic.
+author: Derek Chiang (@derekchiang), Garvit Khatri (@plusminushalf), Fil Makarov (@filmakarov), Kristof Gazso (@kristofgazso), Derek Rein (@arein), Tomas Rocchi (@tomiir), bumblefudge (@bumblefudge)
+discussions-to: https://ethereum-magicians.org/t/erc-7679-smart-account-interfaces/19547
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-04-05
+requires: 4337
+---
+
+## Abstract
+
+Different [ERC-4337](./eip-4337.md) smart account implementations encode their signature, nonce, and calldata differently.  This makes it difficult for DApps, wallets, and smart account toolings to integrate with smart accounts without integrating with account-specific SDKs, which introduces vendor lock-in and hurts smart account adoption.
+
+We propose a standard way for smart account implementations to put their account-specific encoding logic on-chain.
+
+## Motivation
+
+At the moment, to build a [ERC-4337](./eip-4337.md) UserOperation (UserOp for short) for a smart account requires detailed knowledge of how the smart account implementation works, since each implementation is free to encode its nonce, calldata, and signature differently.
+
+As a simple example, one account might use an execution function called `executeFoo`, whereas another account might use an execution function called `executeBar`.  This will result in the `calldata` being different between the two accounts, even if they are executing the same call.
+
+Therefore, someone who wants to send a UserOp for a given smart account needs to:
+
+* Figure out which smart account implementation the account is using.
+* Correctly encode signature/nonce/calldata given the smart account implementation, or use an account-specific SDK that knows how to do that.
+
+In practice, this means that most DApps, wallets, and AA toolings today are tied to a specific smart account implementation, resulting in fragmentation and vendor lock-in.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### UserOp builder
+
+To conform to this standard, a smart account implementation must provide a “UserOp builder” contract that implements the IUserOperationBuilder interface, as defined below:
+
+
+```solidity
+struct Execution {
```

Comment:
Curious why it is termed as "execution" instead of "call"? Feel like we should make terminology symmetrical to [EIP-5792](https://eips.ethereum.org/EIPS/eip-5792).